pkg/snapshotmgr/snapshot_manager.go

```diff
@@ -521,3 +522,101 @@ func (this *SnapshotManager) CreateVolumeFromSnapshot(peID astrolabe.ProtectedEn
 	updatedID, err = astrolabe.NewProtectedEntityIDFromString(download.Status.VolumeID)
 	return
 }
+
+func (this *SnapshotManager) CreateVolumeFromSnapshotWithMetadata(peID astrolabe.ProtectedEntityID, metadata []byte, snapshotID string, backupRepositoryName string) (astrolabe.ProtectedEntityID, *v1api.Download, error) {
+	this.Infof("CreateVolumeFromSnapshotWithMetadata: Start creating restore for %s", peID.String())
+	// TODO(xyang): Will enable following code when astrolabe.Overwrite is ready
+	//config, err := rest.InClusterConfig()
+	//if err != nil {
+	//	this.WithError(err).Errorf("Failed to get k8s inClusterConfig")
+	//	return
+	//}
+	//pluginClient, err := plugin_clientset.NewForConfig(config)
+	//if err != nil {
+	//	this.WithError(err).Errorf("Failed to get k8s clientset with the given config: %v", config)
+	//	return
+	//}
+
+	veleroNs, exist := os.LookupEnv("VELERO_NAMESPACE")
+	if !exist {
+		this.Errorf("CreateVolumeFromSnapshot: Failed to lookup the env variable for velero namespace")
+		return astrolabe.ProtectedEntityID{}, nil, errors.New(fmt.Sprintf("failed to lookup the env variable for velero namespace"))
+	}
+
+	// Retrieve PVC PE TypeManager
+	peTM := this.Pem.GetProtectedEntityTypeManager("pvc")
+	pvcPETM := peTM.(*astrolabe_pvc.PVCProtectedEntityTypeManager)
+	ctx := context.Background()
+	var pe astrolabe.ProtectedEntity
+
+	this.Infof("Ready to call astrolabe CreateFromMetadata API. Will retry on InvalidState error once per second for an hour at maximum")
+	var err error
+	err = wait.PollImmediate(time.Second, time.Hour, func() (bool, error) {
+		// CreateFromMetadata returns a PVCPE
+		pe, err = pvcPETM.CreateFromMetadata(ctx, metadata)
+		if err != nil {
+			if strings.Contains(err.Error(), "The operation is not allowed in the current state") {
+				this.Warnf("Keep retrying on InvalidState error")
+				return false, nil
```

Comment:
Does CreateFromMetadata throw a InvalidState fault? 

Additionally, string comparison is pretty hacky in deciding error type. It might be better to define some error types in astrolabe side if necessary.

Reply:
This is from the Snapshot code above.  Removed it.